src/context/toastNotificationProvider.tsx

```diff
@@ -0,0 +1,252 @@
+import {
+  NotificationAction,
+  NotificationType,
+  ValueOf,
+  failure,
+  info,
+  success,
+} from "@canonical/react-components";
+import { NotificationSeverity } from "@canonical/react-components/dist/components/Notification/Notification";
+import ToastNotification from "components/ToastNotification";
+import ToastNotificationList from "components/ToastNotificationList";
+import React, {
+  FC,
+  PropsWithChildren,
+  ReactNode,
+  createContext,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { Filters } from "util/notifications";
+
+const HIDE_NOTIFICATION_DELAY = 5 * 1000;
+
+export type ToastNotificationType = NotificationType & {
+  timestamp?: ReactNode;
+  id: string;
+};
+
+type ToastNotificationHelper = {
+  notifications: ToastNotificationType[];
+  success: (message: ReactNode) => ToastNotificationType;
+  info: (message: ReactNode, title?: string) => ToastNotificationType;
+  failure: (
+    title: string,
+    error: unknown,
+    message?: ReactNode,
+    actions?: NotificationAction[],
+  ) => ToastNotificationType;
+  clear: (notification: ToastNotificationType) => void;
+  clearAll: () => void;
+  toggleListView: () => void;
+  isListView: boolean;
+  countBySeverity: GroupedNotificationCount;
+};
+
+export type GroupedNotificationCount = {
+  [key in ValueOf<typeof NotificationSeverity>]?: number;
+};
+
+const initialNotification: ToastNotificationType = {
+  id: "",
+  message: "",
+  type: "positive",
+};
+const ToastNotificationContext = createContext<ToastNotificationHelper>({
+  notifications: [],
+  success: () => initialNotification,
+  info: () => initialNotification,
+  failure: () => initialNotification,
+  clear: () => null,
+  clearAll: () => null,
+  toggleListView: () => null,
+  isListView: false,
+  countBySeverity: {},
+});
+
+interface Props {}
+
+const ToastNotificationProvider: FC<PropsWithChildren<Props>> = ({
+  children,
+}) => {
+  const [notifications, setNotifications] = useState<ToastNotificationType[]>(
+    [],
+  );
+  const [countBySeverity, setCountBySeverity] =
+    useState<GroupedNotificationCount>({
+      caution: 0,
+      information: 0,
+      negative: 0,
+      positive: 0,
+    });
```

Comment:
We shouldn't keep the counts and the notifications in state. The counts should be calculated on the fly from the notifications. The problem with keeping both in state is, that the counts and the actual representation in the notifications mismatch. Also it complicates any state updates, because we need to update the notifications and the counts, instead of updating the notifications and relying on the render cycle to calculate the new counts.

Reply:
This was done previously out of consideration for performance. I think I may have attempted to optimise too early in this case. The count is no longer a state and recalculated per render cycle.